File: game_engine.ts
/**
 * Fantasy Loop - Text-Based Game Engine (TypeScript)
 * 
 * A flexible text-based game engine for creating interactive fiction and adventure games.
 */

import * as fs from 'fs';
import * as path from 'path';
import * as readline from 'readline';

interface ItemData {
    name: string;
    description: string;
    usable?: boolean;
    use_message?: string;
}

interface RoomData {
    name: string;
    description: string;
    exits?: { [key: string]: string };
    items?: ItemData[];
    visited?: boolean;
}

interface PlayerData {
    name: string;
    inventory: ItemData[];
    health: number;
    max_health: number;
    current_room: string;
}

interface GameConfig {
    title?: string;
    intro?: string;
    start_room?: string;
    rooms: { [key: string]: RoomData };
}

interface SaveData {
    player: PlayerData;
    rooms: { [key: string]: RoomData };
}

/**
 * Represents an item that can be picked up, used, or examined.
 */
class Item {
    constructor(
        public name: string,
        public description: string,
        public usable: boolean = false,
        public use_message: string = ""
    ) {}

    toJSON(): ItemData {
        return {
            name: this.name,
            description: this.description,
            usable: this.usable,
            use_message: this.use_message
        };
    }

    static fromJSON(data: ItemData): Item {
        return new Item(
            data.name,
            data.description,
            data.usable || false,
            data.use_message || ""
        );
    }
}

/**
 * Represents a location in the game world.
 */
class Room {
    public visited: boolean = false;

    constructor(
        public name: string,
        public description: string,
        public exits: { [key: string]: string } = {},
        public items: Item[] = []
    ) {}

    addExit(direction: string, roomId: string): void {
        this.exits[direction] = roomId;
    }

    addItem(item: Item): void {
        this.items.push(item);
    }

    removeItem(itemName: string): Item | null {
        const index = this.items.findIndex(
            item => item.name.toLowerCase() === itemName.toLowerCase()
        );
        if (index !== -1) {
            return this.items.splice(index, 1)[0];
        }
        return null;
    }

    getItem(itemName: string): Item | null {
        return this.items.find(
            item => item.name.toLowerCase() === itemName.toLowerCase()
        ) || null;
    }

    toJSON(): RoomData {
        return {
            name: this.name,
            description: this.description,
            exits: this.exits,
            items: this.items.map(item => item.toJSON()),
            visited: this.visited
        };
    }
}

/**
 * Represents the player character.
 */
class Player {
    public inventory: Item[] = [];
    public health: number = 100;
    public max_health: number = 100;
    public current_room: string = "start";

    constructor(public name: string = "Hero") {}

    addItem(item: Item): void {
        this.inventory.push(item);
    }

    removeItem(itemName: string): Item | null {
        const index = this.inventory.findIndex(
            item => item.name.toLowerCase() === itemName.toLowerCase()
        );
        if (index !== -1) {
            return this.inventory.splice(index, 1)[0];
        }
        return null;
    }

    hasItem(itemName: string): boolean {
        return this.inventory.some(
            item => item.name.toLowerCase() === itemName.toLowerCase()
        );
    }

    getItem(itemName: string): Item | null {
        return this.inventory.find(
            item => item.name.toLowerCase() === itemName.toLowerCase()
        ) || null;
    }

    toJSON(): PlayerData {
        return {
            name: this.name,
            inventory: this.inventory.map(item => item.toJSON()),
            health: this.health,
            max_health: this.max_health,
            current_room: this.current_room
        };
    }
}

/**
 * Main game engine that manages the game state and logic.
 */
class Game {
    public player: Player = new Player();
    public rooms: Map<string, Room> = new Map();
    public running: boolean = false;
    public game_title: string = "Fantasy Loop";
    public game_intro: string = "Welcome to the adventure!";
    private rl: readline.Interface | null = null;

    addRoom(roomId: string, room: Room): void {
        this.rooms.set(roomId, room);
    }

    getCurrentRoom(): Room | null {
        return this.rooms.get(this.player.current_room) || null;
    }

    movePlayer(direction: string): boolean {
        const currentRoom = this.getCurrentRoom();
        if (!currentRoom) {
            return false;
        }

        const newRoomId = currentRoom.exits[direction.toLowerCase()];
        if (newRoomId) {
            this.player.current_room = newRoomId;
            const newRoom = this.getCurrentRoom();
            if (newRoom) {
                newRoom.visited = true;
            }
            return true;
        }
        return false;
    }

    look(): string {
        const room = this.getCurrentRoom();
        if (!room) {
            return "You are nowhere.";
        }

        let output = `\n${room.name}\n${"=".repeat(room.name.length)}\n${room.description}\n`;

        if (room.items.length > 0) {
            output += "\nYou can see:\n";
            room.items.forEach(item => {
                output += `  - ${item.name}\n`;
            });
        }

        if (Object.keys(room.exits).length > 0) {
            output += "\nExits: " + Object.keys(room.exits).join(", ") + "\n";
        }

        return output;
    }

    takeItem(itemName: string): string {
        const room = this.getCurrentRoom();
        if (!room) {
            return "You are nowhere.";
        }

        const item = room.removeItem(itemName);
        if (item) {
            this.player.addItem(item);
            return `You took the ${item.name}.`;
        }
        return `There is no '${itemName}' here.`;
    }

    dropItem(itemName: string): string {
        const item = this.player.removeItem(itemName);
        if (item) {
            const room = this.getCurrentRoom();
            if (room) {
                room.addItem(item);
                return `You dropped the ${item.name}.`;
            }
        }
        return `You don't have '${itemName}'.`;
    }

    examineItem(itemName: string): string {
        // Check inventory first
        let item = this.player.getItem(itemName);
        if (item) {
            return item.description;
        }

        // Check room
        const room = this.getCurrentRoom();
        if (room) {
            item = room.getItem(itemName);
            if (item) {
                return item.description;
            }
        }

        return `You don't see any '${itemName}' here.`;
    }

    useItem(itemName: string): string {
        const item = this.player.getItem(itemName);
        if (!item) {
            return `You don't have '${itemName}'.`;
        }

        if (!item.usable) {
            return `You can't use the ${item.name}.`;
        }

        return item.use_message;
    }

    showInventory(): string {
        if (this.player.inventory.length === 0) {
            return "Your inventory is empty.";
        }

        let output = "Inventory:\n";
        this.player.inventory.forEach(item => {
            output += `  - ${item.name}\n`;
        });
        return output;
    }

    showStatus(): string {
        return `\n${this.player.name}\nHealth: ${this.player.health}/${this.player.max_health}\n`;
    }

    processCommand(command: string): string {
        command = command.trim().toLowerCase();
        const parts = command.split(/\s+/);

        if (parts.length === 0 || parts[0] === '') {
            return "Please enter a command.";
        }

        const action = parts[0];

        // Movement commands
        if (action === 'go' || action === 'move' || action === 'walk') {
            if (parts.length < 2) {
                return "Go where?";
            }
            const direction = parts[1];
            if (this.movePlayer(direction)) {
                return this.look();
            }
            return "You can't go that way.";
        }

        // Direction shortcuts
        const directions = ['north', 'south', 'east', 'west', 'n', 's', 'e', 'w',
                          'up', 'down', 'northeast', 'northwest', 'southeast', 'southwest'];
        if (directions.includes(action)) {
            if (this.movePlayer(action)) {
                return this.look();
            }
            return "You can't go that way.";
        }

        // Look command
        if (action === 'look' || action === 'l') {
            return this.look();
        }

        // Take/get command
        if (action === 'take' || action === 'get' || action === 'grab' || action === 'pickup') {
            if (parts.length < 2) {
                return "Take what?";
            }
            const itemName = parts.slice(1).join(" ");
            return this.takeItem(itemName);
        }

        // Drop command
        if (action === 'drop' || action === 'put') {
            if (parts.length < 2) {
                return "Drop what?";
            }
            const itemName = parts.slice(1).join(" ");
            return this.dropItem(itemName);
        }

        // Examine command
        if (action === 'examine' || action === 'inspect' || action === 'ex' || action === 'x') {
            if (parts.length < 2) {
                return "Examine what?";
            }
            const itemName = parts.slice(1).join(" ");
            return this.examineItem(itemName);
        }

        // Use command
        if (action === 'use') {
            if (parts.length < 2) {
                return "Use what?";
            }
            const itemName = parts.slice(1).join(" ");
            return this.useItem(itemName);
        }

        // Inventory command
        if (action === 'inventory' || action === 'inv' || action === 'i') {
            return this.showInventory();
        }

        // Status command
        if (action === 'status' || action === 'stats') {
            return this.showStatus();
        }

        // Help command
        if (action === 'help' || action === 'h' || action === '?') {
            return this.showHelp();
        }

        // Quit command
        if (action === 'quit' || action === 'exit' || action === 'q') {
            this.running = false;
            return "Thanks for playing!";
        }

        // Save command
        if (action === 'save') {
            const filename = parts.length > 1 ? parts[1] : "savegame.sav";
            if (this.saveGame(filename)) {
                return `Game saved to ${filename}`;
            }
            return "Failed to save game.";
        }

        // Load command
        if (action === 'load') {
            if (parts.length < 2) {
                return "Load which file?";
            }
            if (this.loadGame(parts[1])) {
                return `Game loaded from ${parts[1]}\n\n${this.look()}`;
            }
            return "Failed to load game.";
        }

        return "I don't understand that command. Type 'help' for a list of commands.";
    }

    showHelp(): string {
        return `
Available Commands:
  Movement: go [direction], north/n, south/s, east/e, west/w
  Actions: take [item], drop [item], examine [item], use [item]
  Info: look/l, inventory/inv/i, status
  System: help/h, save [filename], load [filename], quit/q
`;
    }

    saveGame(filename: string): boolean {
        try {
            const saveDir = "saves";
            if (!fs.existsSync(saveDir)) {
                fs.mkdirSync(saveDir, { recursive: true });
            }
            const filepath = path.join(saveDir, filename);

            const saveData: SaveData = {
                player: this.player.toJSON(),
                rooms: {}
            };

            this.rooms.forEach((room, roomId) => {
                saveData.rooms[roomId] = room.toJSON();
            });

            fs.writeFileSync(filepath, JSON.stringify(saveData, null, 2));
            return true;
        } catch (error) {
            console.error(`Error saving game: ${error}`);
            return false;
        }
    }

    loadGame(filename: string): boolean {
        try {
            const saveDir = "saves";
            const filepath = path.join(saveDir, filename);

            const saveData: SaveData = JSON.parse(fs.readFileSync(filepath, 'utf-8'));

            // Restore player
            const playerData = saveData.player;
            this.player.name = playerData.name;
            this.player.health = playerData.health;
            this.player.max_health = playerData.max_health;
            this.player.current_room = playerData.current_room;
            this.player.inventory = playerData.inventory.map(item => Item.fromJSON(item));

            // Restore rooms (only update visited status and items)
            Object.entries(saveData.rooms).forEach(([roomId, roomData]) => {
                const room = this.rooms.get(roomId);
                if (room) {
                    room.visited = roomData.visited || false;
                    room.items = (roomData.items || []).map(item => Item.fromJSON(item));
                }
            });

            return true;
        } catch (error) {
            console.error(`Error loading game: ${error}`);
            return false;
        }
    }

    loadFromConfig(configFile: string): void {
        try {
            const config: GameConfig = JSON.parse(fs.readFileSync(configFile, 'utf-8'));

            this.game_title = config.title || 'Fantasy Loop';
            this.game_intro = config.intro || 'Welcome to the adventure!';

            // Load rooms
            Object.entries(config.rooms).forEach(([roomId, roomData]) => {
                const items = (roomData.items || []).map(itemData =>
                    new Item(
                        itemData.name,
                        itemData.description,
                        itemData.usable || false,
                        itemData.use_message || ""
                    )
                );

                const room = new Room(
                    roomData.name,
                    roomData.description,
                    roomData.exits || {},
                    items
                );
                this.addRoom(roomId, room);
            });

            // Set starting room
            this.player.current_room = config.start_room || 'start';

            // Mark starting room as visited
            const startRoom = this.getCurrentRoom();
            if (startRoom) {
                startRoom.visited = true;
            }
        } catch (error) {
            console.error(`Error loading configuration: ${error}`);
        }
    }

    async start(): Promise<void> {
        console.log(`\n${"=".repeat(60)}`);
        console.log(this.game_title.padStart((60 + this.game_title.length) / 2));
        console.log(`${"=".repeat(60)}\n`);
        console.log(this.game_intro);
        console.log("\nType 'help' for a list of commands.\n");
        console.log(this.look());

        this.running = true;

        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });

        const promptUser = () => {
            if (!this.running) {
                this.rl?.close();
                return;
            }

            this.rl?.question("\n> ", (command: string) => {
                if (command.trim()) {
                    const result = this.processCommand(command);
                    console.log(result);
                }
                promptUser();
            });
        };

        promptUser();
    }
}

/**
 * Create a demo game for testing.
 */
function createDemoGame(): Game {
    const game = new Game();

    // Create rooms
    const entrance = new Room(
        "Castle Entrance",
        "You stand before a massive stone castle. The entrance gate is open, inviting you inside.",
        { "north": "hall", "east": "garden" }
    );
    entrance.addItem(new Item(
        "sword",
        "A sturdy iron sword, well-balanced and sharp.",
        true,
        "You swing the sword through the air. It feels good in your hand."
    ));

    const hall = new Room(
        "Great Hall",
        "A vast hall with high ceilings and ancient tapestries on the walls. A grand staircase leads upward.",
        { "south": "entrance", "up": "tower", "west": "library" }
    );
    hall.addItem(new Item("key", "A rusty old key with strange markings."));

    const tower = new Room(
        "Tower Chamber",
        "At the top of the tower, you have a magnificent view of the surrounding lands. A treasure chest sits in the corner.",
        { "down": "hall" }
    );
    tower.addItem(new Item("treasure", "A chest filled with glittering gold coins!"));

    const library = new Room(
        "Ancient Library",
        "Shelves upon shelves of dusty old books surround you. The air smells of aged paper.",
        { "east": "hall" }
    );
    library.addItem(new Item(
        "book",
        "An ancient tome with magical symbols on the cover. It radiates a faint blue glow.",
        true,
        "You open the book and feel a surge of magical energy!"
    ));

    const garden = new Room(
        "Castle Garden",
        "A beautiful garden with exotic flowers and a small fountain. The sound of water is soothing.",
        { "west": "entrance" }
    );
    garden.addItem(new Item("flower", "A rare blue rose with an enchanting fragrance."));

    // Add rooms to game
    game.addRoom("entrance", entrance);
    game.addRoom("hall", hall);
    game.addRoom("tower", tower);
    game.addRoom("library", library);
    game.addRoom("garden", garden);

    game.player.current_room = "entrance";
    entrance.visited = true;

    game.game_title = "Fantasy Loop - Demo Adventure";
    game.game_intro = "Welcome, brave adventurer! Your quest begins at the ancient castle.";

    return game;
}

// Main execution
if (require.main === module) {
    (async () => {
        let game: Game;

        if (fs.existsSync("game_config.json")) {
            game = new Game();
            game.loadFromConfig("game_config.json");
        } else {
            game = createDemoGame();
        }

        await game.start();
    })();
}

export { Game, Player, Room, Item, createDemoGame };

File: game_engine.py
"""
Fantasy Loop - Text-Based Game Engine (Python)

A flexible text-based game engine for creating interactive fiction and adventure games.
"""

import json
import os
from typing import Dict, List, Optional, Any


class Item:
    """Represents an item that can be picked up, used, or examined."""
    
    def __init__(self, name: str, description: str, usable: bool = False, use_message: str = ""):
        self.name = name
        self.description = description
        self.usable = usable
        self.use_message = use_message
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert item to dictionary for saving."""
        return {
            'name': self.name,
            'description': self.description,
            'usable': self.usable,
            'use_message': self.use_message
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Item':
        """Create item from dictionary."""
        return cls(
            data['name'],
            data['description'],
            data.get('usable', False),
            data.get('use_message', '')
        )


class Room:
    """Represents a location in the game world."""
    
    def __init__(self, name: str, description: str, exits: Optional[Dict[str, str]] = None, 
                 items: Optional[List[Item]] = None):
        self.name = name
        self.description = description
        self.exits = exits or {}
        self.items = items or []
        self.visited = False
    
    def add_exit(self, direction: str, room_id: str):
        """Add an exit to another room."""
        self.exits[direction] = room_id
    
    def add_item(self, item: Item):
        """Add an item to the room."""
        self.items.append(item)
    
    def remove_item(self, item_name: str) -> Optional[Item]:
        """Remove and return an item from the room."""
        for i, item in enumerate(self.items):
            if item.name.lower() == item_name.lower():
                return self.items.pop(i)
        return None
    
    def get_item(self, item_name: str) -> Optional[Item]:
        """Get an item without removing it."""
        for item in self.items:
            if item.name.lower() == item_name.lower():
                return item
        return None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert room to dictionary for saving."""
        return {
            'name': self.name,
            'description': self.description,
            'exits': self.exits,
            'items': [item.to_dict() for item in self.items],
            'visited': self.visited
        }


class Player:
    """Represents the player character."""
    
    def __init__(self, name: str = "Hero"):
        self.name = name
        self.inventory: List[Item] = []
        self.health = 100
        self.max_health = 100
        self.current_room = "start"
    
    def add_item(self, item: Item):
        """Add an item to inventory."""
        self.inventory.append(item)
    
    def remove_item(self, item_name: str) -> Optional[Item]:
        """Remove and return an item from inventory."""
        for i, item in enumerate(self.inventory):
            if item.name.lower() == item_name.lower():
                return self.inventory.pop(i)
        return None
    
    def has_item(self, item_name: str) -> bool:
        """Check if player has an item."""
        return any(item.name.lower() == item_name.lower() for item in self.inventory)
    
    def get_item(self, item_name: str) -> Optional[Item]:
        """Get an item from inventory."""
        for item in self.inventory:
            if item.name.lower() == item_name.lower():
                return item
        return None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert player to dictionary for saving."""
        return {
            'name': self.name,
            'inventory': [item.to_dict() for item in self.inventory],
            'health': self.health,
            'max_health': self.max_health,
            'current_room': self.current_room
        }


class Game:
    """Main game engine that manages the game state and logic."""
    
    def __init__(self):
        self.player = Player()
        self.rooms: Dict[str, Room] = {}
        self.running = False
        self.game_title = "Fantasy Loop"
        self.game_intro = "Welcome to the adventure!"
    
    def add_room(self, room_id: str, room: Room):
        """Add a room to the game world."""
        self.rooms[room_id] = room
    
    def get_current_room(self) -> Optional[Room]:
        """Get the room the player is currently in."""
        return self.rooms.get(self.player.current_room)
    
    def move_player(self, direction: str) -> bool:
        """Move player in a direction. Returns True if successful."""
        current_room = self.get_current_room()
        if not current_room:
            return False
        
        if direction.lower() in current_room.exits:
            self.player.current_room = current_room.exits[direction.lower()]
            new_room = self.get_current_room()
            if new_room:
                new_room.visited = True
            return True
        return False
    
    def look(self) -> str:
        """Get description of current room."""
        room = self.get_current_room()
        if not room:
            return "You are nowhere."
        
        output = f"\n{room.name}\n{'=' * len(room.name)}\n{room.description}\n"
        
        if room.items:
            output += "\nYou can see:\n"
            for item in room.items:
                output += f"  - {item.name}\n"
        
        if room.exits:
            output += "\nExits: " + ", ".join(room.exits.keys()) + "\n"
        
        return output
    
    def take_item(self, item_name: str) -> str:
        """Take an item from the current room."""
        room = self.get_current_room()
        if not room:
            return "You are nowhere."
        
        item = room.remove_item(item_name)
        if item:
            self.player.add_item(item)
            return f"You took the {item.name}."
        return f"There is no '{item_name}' here."
    
    def drop_item(self, item_name: str) -> str:
        """Drop an item from inventory into current room."""
        item = self.player.remove_item(item_name)
        if item:
            room = self.get_current_room()
            if room:
                room.add_item(item)
                return f"You dropped the {item.name}."
        return f"You don't have '{item_name}'."
    
    def examine_item(self, item_name: str) -> str:
        """Examine an item in inventory or room."""
        # Check inventory first
        item = self.player.get_item(item_name)
        if item:
            return item.description
        
        # Check room
        room = self.get_current_room()
        if room:
            item = room.get_item(item_name)
            if item:
                return item.description
        
        return f"You don't see any '{item_name}' here."
    
    def use_item(self, item_name: str) -> str:
        """Use an item from inventory."""
        item = self.player.get_item(item_name)
        if not item:
            return f"You don't have '{item_name}'."
        
        if not item.usable:
            return f"You can't use the {item.name}."
        
        return item.use_message
    
    def show_inventory(self) -> str:
        """Show player's inventory."""
        if not self.player.inventory:
            return "Your inventory is empty."
        
        output = "Inventory:\n"
        for item in self.player.inventory:
            output += f"  - {item.name}\n"
        return output
    
    def show_status(self) -> str:
        """Show player's status."""
        return f"\n{self.player.name}\nHealth: {self.player.health}/{self.player.max_health}\n"
    
    def process_command(self, command: str) -> str:
        """Process a player command and return the result."""
        command = command.strip().lower()
        parts = command.split()
        
        if not parts:
            return "Please enter a command."
        
        action = parts[0]
        
        # Movement commands
        if action in ['go', 'move', 'walk']:
            if len(parts) < 2:
                return "Go where?"
            direction = parts[1]
            if self.move_player(direction):
                return self.look()
            return "You can't go that way."
        
        # Direction shortcuts
        if action in ['north', 'south', 'east', 'west', 'n', 's', 'e', 'w',
                      'up', 'down', 'northeast', 'northwest', 'southeast', 'southwest']:
            if self.move_player(action):
                return self.look()
            return "You can't go that way."
        
        # Look command
        if action in ['look', 'l']:
            return self.look()
        
        # Take/get command
        if action in ['take', 'get', 'grab', 'pickup']:
            if len(parts) < 2:
                return "Take what?"
            item_name = " ".join(parts[1:])
            return self.take_item(item_name)
        
        # Drop command
        if action in ['drop', 'put']:
            if len(parts) < 2:
                return "Drop what?"
            item_name = " ".join(parts[1:])
            return self.drop_item(item_name)
        
        # Examine command
        if action in ['examine', 'inspect', 'ex', 'x']:
            if len(parts) < 2:
                return "Examine what?"
            item_name = " ".join(parts[1:])
            return self.examine_item(item_name)
        
        # Use command
        if action in ['use']:
            if len(parts) < 2:
                return "Use what?"
            item_name = " ".join(parts[1:])
            return self.use_item(item_name)
        
        # Inventory command
        if action in ['inventory', 'inv', 'i']:
            return self.show_inventory()
        
        # Status command
        if action in ['status', 'stats']:
            return self.show_status()
        
        # Help command
        if action in ['help', 'h', '?']:
            return self.show_help()
        
        # Quit command
        if action in ['quit', 'exit', 'q']:
            self.running = False
            return "Thanks for playing!"
        
        # Save command
        if action in ['save']:
            filename = parts[1] if len(parts) > 1 else "savegame.sav"
            if self.save_game(filename):
                return f"Game saved to {filename}"
            return "Failed to save game."
        
        # Load command
        if action in ['load']:
            if len(parts) < 2:
                return "Load which file?"
            if self.load_game(parts[1]):
                return f"Game loaded from {parts[1]}\n\n" + self.look()
            return "Failed to load game."
        
        return "I don't understand that command. Type 'help' for a list of commands."
    
    def show_help(self) -> str:
        """Show help text."""
        return """
Available Commands:
  Movement: go [direction], north/n, south/s, east/e, west/w
  Actions: take [item], drop [item], examine [item], use [item]
  Info: look/l, inventory/inv/i, status
  System: help/h, save [filename], load [filename], quit/q
"""
    
    def save_game(self, filename: str) -> bool:
        """Save the game state to a file."""
        try:
            save_dir = "saves"
            os.makedirs(save_dir, exist_ok=True)
            filepath = os.path.join(save_dir, filename)
            
            save_data = {
                'player': self.player.to_dict(),
                'rooms': {room_id: room.to_dict() for room_id, room in self.rooms.items()}
            }
            
            with open(filepath, 'w') as f:
                json.dump(save_data, f, indent=2)
            return True
        except Exception as e:
            print(f"Error saving game: {e}")
            return False
    
    def load_game(self, filename: str) -> bool:
        """Load the game state from a file."""
        try:
            save_dir = "saves"
            filepath = os.path.join(save_dir, filename)
            
            with open(filepath, 'r') as f:
                save_data = json.load(f)
            
            # Restore player
            player_data = save_data['player']
            self.player.name = player_data['name']
            self.player.health = player_data['health']
            self.player.max_health = player_data['max_health']
            self.player.current_room = player_data['current_room']
            self.player.inventory = [Item.from_dict(item) for item in player_data['inventory']]
            
            # Restore rooms (only update visited status and items)
            for room_id, room_data in save_data['rooms'].items():
                if room_id in self.rooms:
                    self.rooms[room_id].visited = room_data['visited']
                    self.rooms[room_id].items = [Item.from_dict(item) for item in room_data['items']]
            
            return True
        except Exception as e:
            print(f"Error loading game: {e}")
            return False
    
    def load_from_config(self, config_file: str):
        """Load game configuration from a JSON file."""
        try:
            with open(config_file, 'r') as f:
                config = json.load(f)
            
            self.game_title = config.get('title', 'Fantasy Loop')
            self.game_intro = config.get('intro', 'Welcome to the adventure!')
            
            # Load rooms
            for room_id, room_data in config.get('rooms', {}).items():
                items = [Item(
                    item['name'],
                    item['description'],
                    item.get('usable', False),
                    item.get('use_message', '')
                ) for item in room_data.get('items', [])]
                
                room = Room(
                    room_data['name'],
                    room_data['description'],
                    room_data.get('exits', {}),
                    items
                )
                self.add_room(room_id, room)
            
            # Set starting room
            self.player.current_room = config.get('start_room', 'start')
            
            # Mark starting room as visited
            start_room = self.get_current_room()
            if start_room:
                start_room.visited = True
            
        except Exception as e:
            print(f"Error loading configuration: {e}")
    
    def start(self):
        """Start the game loop."""
        print(f"\n{'=' * 60}")
        print(f"{self.game_title.center(60)}")
        print(f"{'=' * 60}\n")
        print(self.game_intro)
        print("\nType 'help' for a list of commands.\n")
        print(self.look())
        
        self.running = True
        
        while self.running:
            try:
                command = input("\n> ").strip()
                if command:
                    result = self.process_command(command)
                    print(result)
            except KeyboardInterrupt:
                print("\n\nGame interrupted.")
                self.running = False
            except EOFError:
                print("\n\nGame ended.")
                self.running = False


def create_demo_game() -> Game:
    """Create a demo game for testing."""
    game = Game()
    
    # Create rooms
    entrance = Room(
        "Castle Entrance",
        "You stand before a massive stone castle. The entrance gate is open, inviting you inside.",
        {"north": "hall", "east": "garden"}
    )
    entrance.add_item(Item("sword", "A sturdy iron sword, well-balanced and sharp.", True, 
                           "You swing the sword through the air. It feels good in your hand."))
    
    hall = Room(
        "Great Hall",
        "A vast hall with high ceilings and ancient tapestries on the walls. " +
        "A grand staircase leads upward.",
        {"south": "entrance", "up": "tower", "west": "library"}
    )
    hall.add_item(Item("key", "A rusty old key with strange markings.", False))
    
    tower = Room(
        "Tower Chamber",
        "At the top of the tower, you have a magnificent view of the surrounding lands. " +
        "A treasure chest sits in the corner.",
        {"down": "hall"}
    )
    tower.add_item(Item("treasure", "A chest filled with glittering gold coins!", False))
    
    library = Room(
        "Ancient Library",
        "Shelves upon shelves of dusty old books surround you. The air smells of aged paper.",
        {"east": "hall"}
    )
    library.add_item(Item("book", "An ancient tome with magical symbols on the cover. " +
                          "It radiates a faint blue glow.", True,
                          "You open the book and feel a surge of magical energy!"))
    
    garden = Room(
        "Castle Garden",
        "A beautiful garden with exotic flowers and a small fountain. " +
        "The sound of water is soothing.",
        {"west": "entrance"}
    )
    garden.add_item(Item("flower", "A rare blue rose with an enchanting fragrance.", False))
    
    # Add rooms to game
    game.add_room("entrance", entrance)
    game.add_room("hall", hall)
    game.add_room("tower", tower)
    game.add_room("library", library)
    game.add_room("garden", garden)
    
    game.player.current_room = "entrance"
    entrance.visited = True
    
    game.game_title = "Fantasy Loop - Demo Adventure"
    game.game_intro = "Welcome, brave adventurer! Your quest begins at the ancient castle."
    
    return game


if __name__ == "__main__":
    # Check if there's a config file
    if os.path.exists("game_config.json"):
        game = Game()
        game.load_from_config("game_config.json")
    else:
        game = create_demo_game()
    
    game.start()
